lisa/tools/virtualclient.py

```diff
@@ -0,0 +1,402 @@
+        for metri_messages in metric_messages:
+            notifier.notify(metri_messages)
+
+    def _parse_timestamp(self, timestamp_str: str) -> datetime:
+        # Strip 'Z' and limit microseconds to 6 digits
+        timestamp_str = timestamp_str.rstrip("Z")
+        # Split timestamp and microseconds
+        if "." in timestamp_str:
+            main_time, microseconds = timestamp_str.split(".")
+            # Truncate microseconds to 6 digits
+            microseconds = microseconds[:6]
+            timestamp_str = f"{main_time}.{microseconds}"
+        return datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%M:%S.%f")
+
+
+class VcRunner:
+    # [07/08/2024 13:58:15] Exit Code: 0\r\n
+    exit_code_pattern = re.compile(r"([\w\W]*?)Exit Code: (?P<exit_code>\d+)")
+
+    def __init__(self, environment: "Environment", roles: List[str]):
+        if len(roles) > len(environment.nodes):
+            raise ValueError(
+                "More roles specified than available nodes in the environment"
+            )
+
+        self._targets: List[VcTargetInfo] = []
+        nodes = [cast("RemoteNode", environment.nodes[i]) for i in range(len(roles))]
+
+        for role, node in zip(roles, nodes):
+            self._targets.append(VcTargetInfo(role=role, node=node))
+
+        self._virtual_clients: List[Any] = run_in_parallel(
+            [
+                partial(target.node.tools.get, VirtualClientTool)
+                for target in self._targets
+            ]
+        )
+
+    def run(
+        self,
+        profile_name: str,
+        experiment_id: str = "",
+        system: str = "Azure",
+        timeout: int = 10,
+    ) -> None:
+        layout_file = self._generate_layout_file()
+        experiment_id = experiment_id or str(uuid.uuid4())
+
+        client_params = self._generate_client_params(
+            profile_name,
+            system,
+            layout_file,
+            timeout,
+            experiment_id,
+        )
+
+        results = self._execute_commands(client_params)
+        self._wait_for_client_result(results, timeout)
+        self._process_results()
+
+    def _generate_layout_file(
+        self,
+    ) -> str:
+        layout_file = ""
+        if len(self._virtual_clients) > 1:
+            for vc in self._virtual_clients:
+                layout_file = vc.generate_layout(self._targets)
+        return layout_file
+
+    def _generate_client_params(
+        self,
+        profile_name: str,
+        system: str,
+        layout_file: str,
+        timeout: int,
+        experiment_id: str,
+    ) -> Any:
+        return {
+            node_info.node.name: {
+                "profile_name": profile_name,
+                "system": system,
+                "client_id": node_info.node.name,
+                "layout_path": layout_file,
+                "timeout": timeout,
+                "experiment_id": experiment_id,
+            }
+            for node_info in self._targets
```

Comment:
As above parameter, when save vc tools into the vc targets, it doesn't need a loop to get all parameters at the same time.